invisible_cities/reco/xy_algorithm_test.py

```diff
@@ -1,106 +1,218 @@
+    y          =              y.astype(np.float)
+
+    return pd.DataFrame(dict( SensorID  =   sensorid,
+                             ChannelID  =  channelid,
+                             Active     =     active,
+                             X          =          x,
+                             Y          =          y,
+                             adc_to_pes = adc_to_pes,
+                             Sigma      =      sigma))
+
+@fixture(scope="session")
+def toy_sipm_signal_and_inds():
+    k   = 10000
+    xs  = np.arange(k)
+    ys  = xs + k
+    pos = np.stack((xs, ys), axis=1)
+    qs  = xs + 2*k
+    i   = np.array([0, 5, 1000, 9999])
+    return k, i, pos, qs
+
+
+@fixture(scope="session")
+def datasipm():
+    return DataSiPM(0)
+
+
+@fixture(scope="session")
+def datasipm_all_active():
+    return DataSiPM(1)
+
+
+@fixture(scope="session")
+def datasipm_5000():
+    return DataSiPM(5000)
+
+
+@fixture(scope="session")
+def datasipm_3x5():
+    # Create fake database with this 3 x 5 grid of SiPMs:
+    # o = active, x = masked
+    #
+    # x - - - >
+    # y | o o o
+    #   | o o x
+    #   | o o o
+    #   | o o o
+    #   v o o o
+    x         = np.tile  (np.arange(3), 5)
+    y         = np.repeat(np.arange(5), 3)
+    active    = np.ones  (15, dtype=bool)
+    active[5] = False
+    return _create_fake_datasipm(x, y, active)
+
+
+@fixture(scope="session")
+def datasipm5x5():
+    # Create fake database with this 5 x 5 grid of SiPMs:
+    # o = active, x = masked
+    #
+    # o o o o o
+    # o x x x o
+    # o x o x o
+    # o x x x o
+    # o o o o o
+    active           = np.ones  ((5, 5), dtype=bool)
+    active[1:4, 1:4] = False
+    active[  2,   2] = True
+    x                = np.tile  (np.arange(5), 5)
+    y                = np.repeat(np.arange(5), 5)
+    return _create_fake_datasipm(x, y, active.flatten())
+
+
+@given(positions_and_qs())
+@settings(max_examples=100)
+def test_barycenter_generic(p_q):
     pos, qs = p_q
     B  = barycenter(pos, qs)[0]
     assert np.allclose(B.posxy, np.average(pos, weights=qs, axis=0))
     assert np. isclose(B.Q  , qs.sum())
     assert B.nsipm == len(qs)
 
 
-def test_barycenter_raises_sipm_empty_list():
-    with raises(SipmEmptyList):
-        barycenter(np.array([]), None)
+@parametrize("     x         y        q     expected_xy".split(),
+             (( (-1, +1), ( 0,  0), (1, 1),   ( 0, 0)  ),
+              ( ( 0,  0), (-1, +1), (1, 1),   ( 0, 0)  ),
+              ( (-1, +1), (-1, +1), (1, 1),   ( 0, 0)  ),
+              ( (-4, +3), (+2, -5), (2, 5),   ( 1,-3)  )))
+def test_barycenter_simple_cases(x, y, q, expected_xy):
+    xy = np.stack((x, y), axis=1)
+    qs = np.array(q)
+    b  = barycenter(xy, qs)[0]
+    assert np.allclose(b.posxy, expected_xy)
 
 
-def test_barycenter_raises_sipm_zero_charge():
-    with raises(SipmZeroCharge):
-        barycenter(np.array([1,2]), np.array([0,0]))
+@given(float_arrays(size=8, min_value=0, max_value=100, mask=np.sum))
```

Comment:
> What's the purpose of the 'np.sum' filter here?

`np.sum` will return the sum of the values generated by the strategies. This will be treated as a boolean, i.e., `False if sum == 0` and `True` otherwise. This is to ensure the strategy doesn't generate an array filled with zeros.

> why not add it to the test_barycenter_generic test?

Simpler tests can help to pinpoint the error in case of failure If `test_barycenter_generic` fails, it can be for a number of reasons. If this one fails, it is because it is returning more than one cluster.

> As in the previous comment, why do you need a generic and a specific case?

The specific case helps to illustrate the workings of the function, the generic case tries to be more robust than just trying a few cases by hand.

Reply:
Ok, I see the reasoning, it's fine with me!